### PR TITLE
Fix issue in parsing tuple type

### LINKED ISSUE #28347: Compiler crashes for a tuple type descriptor with just a rest descriptor when the rest descriptor is another tuple type descriptor
**Description:**
$title.

**Steps to reproduce:**
```ballerina
public function main() {
    [string, [int, int]...] t1 = ["foo", [1, 2]]; // works
    [int[]...] t2 = [[1, 2]]; // works
    [[int, int]...] t3 = [[1, 2]]; // crashes
}
```

```log
[2021-02-01 17:16:22,096] SEVERE {b7a.log.crash} - Node not supported: TypedBindingPatternNode 
java.lang.RuntimeException: Node not supported: TypedBindingPatternNode
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transformSyntaxNode(BLangNodeTransformer.java:3727)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transformSyntaxNode(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.NodeTransformer.transform(NodeTransformer.java:584)
	at io.ballerina.compiler.syntax.tree.TypedBindingPatternNode.apply(TypedBindingPatternNode.java:50)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5052)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:959)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.TupleTypeDescriptorNode.apply(TupleTypeDescriptorNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5052)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createBLangVarDef(BLangNodeTransformer.java:2743)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2700)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.VariableDeclarationNode.apply(VariableDeclarationNode.java:67)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.generateAndAddBLangStatements(BLangNodeTransformer.java:4572)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1483)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.FunctionBodyBlockNode.apply(FunctionBodyBlockNode.java:59)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.populateFunctionNode(BLangNodeTransformer.java:1391)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createFunctionNode(BLangNodeTransformer.java:1371)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1349)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.FunctionDefinitionNode.apply(FunctionDefinitionNode.java:71)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:569)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:492)
	at io.ballerina.projects.DocumentContext.compilationUnit(DocumentContext.java:111)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:355)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleCompilationState$1.generatePlatformSpecificCode(ModuleCompilationState.java:53)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:326)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:136)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:122)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:98)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:135)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:97)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:69)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:155)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
```

**Affected Versions:**
slalpha1

### LINKED ISSUE #28865: Compiler crashes when a tuple type only has a tuple type rest descriptor
**Description:**
$title

**Steps to reproduce:**
```
[[int, int]...] x = [[1, 2], [3, 4], [7, 8]];
```
```
[2021-02-24 12:24:03,272] SEVERE {b7a.log.crash} - Node not supported: TypedBindingPatternNode 
java.lang.RuntimeException: Node not supported: TypedBindingPatternNode
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transformSyntaxNode(BLangNodeTransformer.java:3753)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transformSyntaxNode(BLangNodeTransformer.java:470)
	at io.ballerina.compiler.syntax.tree.NodeTransformer.transform(NodeTransformer.java:584)
	at io.ballerina.compiler.syntax.tree.TypedBindingPatternNode.apply(TypedBindingPatternNode.java:50)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5307)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:977)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:470)
	at io.ballerina.compiler.syntax.tree.TupleTypeDescriptorNode.apply(TupleTypeDescriptorNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createTypeNode(BLangNodeTransformer.java:5307)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createBLangVarDef(BLangNodeTransformer.java:2762)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2719)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:470)
	at io.ballerina.compiler.syntax.tree.VariableDeclarationNode.apply(VariableDeclarationNode.java:67)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.generateAndAddBLangStatements(BLangNodeTransformer.java:4827)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1501)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:470)
	at io.ballerina.compiler.syntax.tree.FunctionBodyBlockNode.apply(FunctionBodyBlockNode.java:59)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.populateFunctionNode(BLangNodeTransformer.java:1409)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createFunctionNode(BLangNodeTransformer.java:1389)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1367)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:470)
	at io.ballerina.compiler.syntax.tree.FunctionDefinitionNode.apply(FunctionDefinitionNode.java:71)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:587)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:470)
	at io.ballerina.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:500)
	at io.ballerina.projects.DocumentContext.compilationUnit(DocumentContext.java:111)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:356)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleCompilationState$1.generatePlatformSpecificCode(ModuleCompilationState.java:53)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:327)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:136)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:122)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:98)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:129)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:97)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:69)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:155)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
```
**Affected Versions:**
slalpha2

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by rdulmina**
## Purpose
Fix issue in parsing tuple type in statement-level where the ambigeous member is also a tuple type.

Fixes #28347

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -16028,13 +16028,14 @@ private STNode parseAsTupleTypeDesc(STNode annots, STNode openBracket, List<STNo
         endContext();
 
         STNode tupleType = STNodeFactory.createTupleTypeDescriptorNode(openBracket, tupleTypeMembers, closeBracket);
+        if (!isRoot) {
+            return tupleType;
+        }
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:16033]: Shouldn't we return here after parsing the complex type desc? Also after ending the context.
At this point, we have two `STMT_START_BRACKETED_LIST` contexts in the stack. Inner one should be terminated before exit ryt?

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: Fixed

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -16059,11 +16059,13 @@ private STNode parseAsTupleTypeDesc(STNode annots, STNode openBracket, List<STNo
         STNode typeDesc =
                 parseComplexTypeDescriptor(tupleType, ParserRuleContext.TYPE_DESC_IN_TYPE_BINDING_PATTERN, true);
         endContext();
-        STNode typedBindingPattern = parseTypedBindingPatternTypeRhs(typeDesc, ParserRuleContext.VAR_DECL_STMT, isRoot);
+
         if (!isRoot) {
-            return typedBindingPattern;
+            return tupleType;
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:16064]:
```suggestion
            return typeDesc;
```

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: thanx :grin:.

**Review by lochana-chathura — APPROVED** [review]

**Pull request merged**